12 build using resources

JacobBredin opened the pull request:
## Summary
Added functionality for building structures, and a new simple behaviour for pawns to do this task. Methods have also been added to the MatrixUtils class.

![build-using-resources-3](https://user-images.githubusercontent.com/70701264/137294029-dc5d0422-3163-417e-8ca0-a7b051162e0e.gif)

## Test Coverage
![image](https://user-images.githubusercontent.com/70701264/137292845-cf4f6496-f2ff-42b2-8f69-c9bf81ff00be.png)

![image](https://user-images.githubusercontent.com/70701264/137292935-f88f6a09-8668-4b2a-bddd-3ebf7cc92a19.png)

![image](https://user-images.githubusercontent.com/70701264/137292996-9cd28a39-389b-40b1-8073-62e8f0afebc3.png)

![image](https://user-images.githubusercontent.com/70701264/137293063-3e21bd1c-8053-4612-bb39-30a8c918fca5.png)

![image](https://user-images.githubusercontent.com/70701264/137293093-bc9cfce9-634e-4d37-8fde-c53587130870.png)

![image](https://user-images.githubusercontent.com/70701264/137293139-57d113ab-b679-4292-9434-3fcf55e54c8b.png)


## Definition of Done

- [x] Completed all Tasks of the User Story
- [x] The Tasks have tests and they all pass
- [x] Overall test coverage is >= 90%
- [x] A UML class diagram of the newly implemented User Story is done and included in the README.md of the package
## Diagrams
### com.thebois.models.structures
![com thebois models world structures](https://user-images.githubusercontent.com/70701264/135406699-b8206ae0-2454-47f3-abbd-9b76f8e02765.jpg)
### com.thebois.models.beings
![com thebois models beings](https://user-images.githubusercontent.com/70701264/135406687-e98cfafa-8c3d-4c04-b8b3-a0c7594ba759.jpg)
### com.thebois.models.world
![com thebois models world](https://user-images.githubusercontent.com/70701264/135406697-bf53dfd7-dd98-4c9c-baad-2556e2c705bb.jpg)
### com.thebois.utils
![com thebois utils](https://user-images.githubusercontent.com/70701264/135406704-21b5fe3a-a8fc-4900-b011-736bc3fe7880.jpg)

Closes #12

JacobBredin committed:
Add IBuildable interface and IItem placeholder

JacobBredin committed:
Merge remote-tracking branch 'origin/master' into 12-build-using-resources

JacobBredin committed:
Merge remote-tracking branch 'origin/master' into 12-build-using-resources

JacobBredin committed:
Make AbstractStructure implement IBuildable

JacobBredin committed:
Refactor method builtStatus in AbstractStructure

JacobBredin committed:
Add House color rendition based on builtStatus

JacobBredin committed:
Add matrixToCollection method to MatrixUtils

JacobBredin committed:
Add a method for informing IBeings of IStructures

JacobBredin committed:
Add update of StructureView in its Controller

JacobBredin committed:
Add methods to IFinder

JacobBredin committed:
Refactor matrix spiral search to seperate method

JacobBredin committed:
Move spiral search method to MatrixUtils

JacobBredin committed:
Add matrixSpiralSearch test to MatrixUtilsTests

JacobBredin committed:
Add IFinder parameter to ABeing and ABeingGroup

JacobBredin committed:
Refactor Pawns to use IFinder

JacobBredin committed:
Add getSubMatrix method to MatrixUtils

JacobBredin committed:
Add test for getSubMatrix

JacobBredin committed:
Fix getSubMatrix bug and add new test arguments

JacobBredin committed:
Remove references and use of updateKnownStructure

JacobBredin committed:
Add readme file and UML diagram to world package

JacobBredin committed:
Remove depricated method findClosetStructure

JacobBredin committed:
Fix mix-up of  X, Y and Row, Col in World

JacobBredin committed:
Add tests to WorldTests

JacobBredin committed:
Add tests to BeingTests

JacobBredin committed:
Change matrixSpiralSearch to return Collection

JacobBredin committed:
Add behaviour to pawns to make them build houses

JacobBredin committed:
Move pawn specific methods to pawn

JacobBredin committed:
Add tests for Pawn and House

JacobBredin committed:
Add builtStatus test

JacobBredin committed:
Add Javadoc and code comments

JacobBredin committed:
Update images in readme files

JacobBredin committed:
Fix CheckStyle errors

martinjonsson01 commented:
Before I review the code itself, your test coverage is low! Could you test the remaining branches in `com.thebois.model.beings` and `com.thebois.models.structures`?

JacobBredin commented:
> Before I review the code itself, your test coverage is low! Could you test the remaining branches in `com.thebois.model.beings` and `com.thebois.models.structures`?

I can't test parts of the code because a lack of IItems, but I know the code works because I've tested it elsewhere.  Should I wait until Colony Inventory is merged?

martinjonsson01 commented:
>I can't test parts of the code because a lack of IItems

There aren't any items yet, so you should've implemented the functionality without requiring items, right? Instead of needing items to construct a building, it's free to construct. Until items are implemented.

Actually, can we consider this user story done if it's not completely implemented yet? You're blocked by #40 but that just needs to be merged and then you can complete the story and create a PR then, right?

> I know the code works because I've tested it elsewhere

You couldn't have, since it would have been counted as covered if you've got tests covering it.

JacobBredin commented:
> There aren't any items yet, so you should've implemented the functionality without requiring items, right? Instead of needing items to construct a building, it's free to construct. Until items are implemented.

I feel like it would be a waist of time to remove code that is currently using IItem, should I just standby for Colony Inventory to be merged?

martinjonsson01 commented:
>I feel like it would be a waist of time to remove code that is currently using IItem

Yeah, that's true. You probably should've not used items at all until they were implemented, but that's easy to say in hindsight. 

>should I just standby for Colony Inventory to be merged?

Yup, that's probably the way to go right now. And I think it's on the edge of being merged so either today or Friday you'll be able to continue, I believe. (Either way we've got documentation to write right now instead of working on user stories :P )

>Closed PR

Actually, we can leave it up if you want, so you can get feedback on what you've written so far. We can convert this PR into a draft PR and then make it into an actual PR once you're completely finished!

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
     }
 
+    @Test
+    public void pawnBuildsHouses() {
```
Generally it's helpful if the test names follow the format:
`whenSomethingIsDoneAndTheObjectIsInThisStateThenThisHappens`
On a lot of these tests, you've only got the:
`ThisHappens`, you never describe what is being done and what state the object is in.

Now, this is easy to see if I read the code of the test, but from the test runner all you can see is the name and whether it failed or not, so it's much more helpful to read that "`updateDeliversItemsToStructureWhenNearby` has failed" rather than
"`pawnBuildsHouses` has failed", if that makes sense?

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+        final Pawn pawn = new Pawn(position, position, Mockito.mock(Random.class), mockFinder);
+
+        // Act
+        final float initialBuildStatus = structure.builtStatus();
+        pawn.update();
+        final float finalBuildStatus = structure.builtStatus();
```
Here you're testing more than just `Pawn`. To isolate one from the other you should mock `IStructure` instead of use a real one, and then assert that the `Pawn` calls the `deliverItem` method of the mocked structure:

```java
// Arrange
[...]
IStructure mockStructure = Mockito.mock(IStructure.class);

// Act
pawn.update();

// Assert
verify(mockStructure, times(1)).deliverItem(any());
```

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+        final Pawn pawn = new Pawn(position, position, Mockito.mock(Random.class), mockFinder);
+
+        // Act
+        final float initialBuildStatus = structure.builtStatus();
+        pawn.update();
+        final float finalBuildStatus = structure.builtStatus();
```
Use a mocked structure and verify that `deliverItems` is called 0 times.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+        final float finalBuildStatus = structure.builtStatus();
+
+        // Assert
+        // Value did not change because Pawn was not close enough
```
This comment right here is what should be obvious from the name of the method, so when you've got a correctly named test then you shouldn't have to explain why you're asserting something! It's a helpful rule of thumb.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+    }
+
+    @Test
+    public void doesPawnWalkToNearestUnBuiltBuilding() {
```
This test name is formulated like a question, but in it you're asserting things. It's a bit contradictory, so a better name would be the more assertive `updateSetsDestinationToClosestUnbuiltStructureWhenThereAreMultipleNearby`.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+    @Test
+    public void doesPawnWalkToNearestUnBuiltBuilding() {
+        // Arrange
+        final Position positionA = new Position(1, 0);
+        final Position positionB = new Position(2, 0);
+        final Position positionC = new Position(3, 0);
```
These names explain nothing about what they represent. 

Is `positionA` the starting position of one of the structures, or is it where the being is located? I don't know because the names are meaningless.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+
+        final IFinder mockFinder = Mockito.mock(IFinder.class);
+
+        when(mockFinder.findNearestStructures(any(), anyInt())).thenReturn(structures);
+        // builtStructure would be found first by the current finder
+        when(mockFinder.findNearestStructure(any())).thenReturn(Optional.of(builtStructure));
```
question

Does `Pawn` use both `findNearestStructures` and `findNearestStructure`, and if so, why use both?

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -160,4 +183,79 @@ public void equalsReturnsFalseForNotEqualBeings(final AbstractBeing first,
+        pawn.update();
+
+        // Assert
+        // Value did not change because Pawn was not close enough
```
This comment does not make any sense, is it outdated? You shouldn't have comments explain what the code does if the code itself explains it the same way.

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -177,4 +165,96 @@ private boolean matrixEquals(final ITile[][] worldMatrix1, final ITile[][] world
+
+    @ParameterizedTest
+    @MethodSource("getCorrectCoordinatesToTest")
+    public void testFindNearestStructure(final int startX,
```
No point in having `test` in the name of a test. It's already apparent that it is in fact a test.

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -177,4 +165,96 @@ private boolean matrixEquals(final ITile[][] worldMatrix1, final ITile[][] world
+            startY));
+
+        // Assert
+        if (foundStructure.isPresent()) {
+            assertThat(foundStructure.get().getPosition()).isEqualTo(new Position(endX, endY));
+        }
+        else {
+            assertThat(foundStructure.isEmpty()).isEqualTo(false);
+        }
```
You shouldn't have conditional assertions, because you arranged the test so you already know what the results should be.

`Optional` has a method called `orElseThrow` that gets the value, and if it isn't present it throws an exception (thus failing the test). 

If you are in fact using a dataset that contains two different things to test, when there is a found structure and when there isn't one, then you've got two tests there. Not one.

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -177,4 +165,96 @@ private boolean matrixEquals(final ITile[][] worldMatrix1, final ITile[][] world
+        final Optional<IStructure> foundStructure = world.findNearestStructure(new Position(
+            startX, startY));
+
+        // Assert
+        if (foundStructure.isPresent()) {
+            assertThat(foundStructure.get().getPosition()).isNotEqualTo(new Position(endX, endY));
+        }
+        else {
+            assertThat(foundStructure.isEmpty()).isEqualTo(true);
+        }
```
This is also probably a sign that you've got two tests in one.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        final IItem item = new IItem() {
+        };
+
+        // Act
+
+        // Assert
+        assertThat(house.deliverItem(item)).isTrue();
```
The acting is the `house.deliverItem(item)`. The asserting is the `assertThat(itemDelivered).isTrue();`

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        house.deliverItem(item);
+
+        // Assert
+        assertThat(house.builtStatus()).isEqualTo(0.01f);
```
There's no way for the test to know what the status will go up by 0.01f. How does it know? Because you know how its implemented in the actual class. This means the test is relying on implementation details, which it shouldn't. Instead, you can just check that the status goes up.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        house.setAllNeededItems(neededItems);
+        house.setDeliveredItems(delivered);
+
+        // Act
+
+        // Assert
+        assertThat(house.neededItems()).isEqualTo(expectedNeededItems);
```
You should separate your acting from your asserting.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+    public void neededItemsNotEqualTest() {
+        // Arrange
+        final House house = new House(new Position());
+        final IItem mockItem1 = Mockito.mock(IItem.class);
+        final IItem mockItem2 = Mockito.mock(IItem.class);
+        final IItem mockItem3 = Mockito.mock(IItem.class);
```
I mentioned in another test that `positionA` `positionB` `positionC` are bad names that make it hard to understand the test.

In this case, it is completely okay to call them `mockItem1` and so on. Why? Because I don't need to know which item is which to understand the test. It doesn't matter because I just need to know that one of the items is delivered and the other two should still be needed.

In the other test with the badly named positions, I needed to know which position was which in order to comprehend how the test worked, which made it really difficult to understand.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        final Collection<IItem> delivered = List.of(mockItem1);
+        final Collection<IItem> expectedNeededItems = List.of(mockItem1);
+
+        house.setAllNeededItems(neededItems);
+        house.setDeliveredItems(delivered);
```
Using a public setters is a code smell, you should use the constructor and the `deliverItem` method.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        house.setAllNeededItems(neededItems);
+        house.setDeliveredItems(delivered);
+
+        // Act
+
+        // Assert
+        assertThat(house.neededItems()).isNotEqualTo(expectedNeededItems);
```
Separate your acting from your asserting.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        house.setAllNeededItems(items);
+        house.setDeliveredItems(items);
+
+        // Act
+
+        // Assert
+        assertThat(house.dismantle(item)).isTrue();
```
Separate

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,4 +39,106 @@ public void getTypeEqualsHouse() {
+        final Collection<IItem> items = new ArrayList<>();
+        items.add(item);
+
+        // Act
+
+        // Assert
+        assertThat(house.dismantle(item)).isFalse();
+    }
```
Separate

martinjonsson01 commented on src/test/java/com/thebois/utils/MatrixUtilsTests.java:
```diff
@@ -44,4 +62,96 @@ public void forEachElementGetsAllElementsInOrder(final Integer[][] matrix,
+            3, 1).iterator().next();
+
+        // Assert
+        if (foundStructure.isPresent()) {
+            assertThat(foundStructure.get()).isEqualTo(expectedString);
+        }
+        else {
+            assertThat(foundStructure.isEmpty()).isEqualTo(false);
+        }
```
Two tests in one.

martinjonsson01 commented on src/test/java/com/thebois/utils/MatrixUtilsTests.java:
```diff
@@ -44,4 +62,96 @@ public void forEachElementGetsAllElementsInOrder(final Integer[][] matrix,
+                                         final int startY,
+                                         final String expectedString) {
+
+        // Arrange
+
```
Looks like you're arranging completely in the parameter method source, which is okay. You don't need to have empty comments like this.

martinjonsson01 commented on src/test/java/com/thebois/utils/MatrixUtilsTests.java:
```diff
@@ -44,4 +62,96 @@ public void forEachElementGetsAllElementsInOrder(final Integer[][] matrix,
+                                 final int startCol,
+                                 final int endRow,
+                                 final int endCol) {
+        // Arrange
+
```
No need for empty comments.

martinjonsson01 reviewed (changes requested):
Reviewing what you've got so far so that you can implement the changes while you're blocked, and so that when you continue working on the story you'll be able to follow the tips.

I've left a bunch of tips and a few questions. If there's anything you want clarification on, or don't agree with, that isn't quick to formulate in text, we could discuss it tomorrow?

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -38,4 +49,29 @@ private void setRandomDestination() {
+    protected void pickNearestIncompleteStructureAsDestination() {
+        final Collection<Optional<IStructure>> structures =
+            getFinder().findNearestStructures(this.getPosition(), 10);
+        final Boolean[] found = {false};
```
Might be missing something obvious but why the array? Seems like we only set the first element to either true or false?

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -74,7 +76,7 @@ public World(final int worldSize, final Colony colony) {
     }
 
     private void initColony() {
-        colony = new Colony(findEmptyPositions(pawnCount));
+        colony = new Colony(findEmptyPositions(pawnCount), this);
```
Feels like this creates some sort of circular dependency. In world we have a constructor that can take colony as an parameter but colony now obviously have a constructor that have world as parameter. Seems like a code smell to me

martinjonsson01 replied:
Note that it's implicitly casting `World` into `IFinder`, so it's not a double dependency at least!

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -188,12 +181,66 @@ public IRoleAllocator getRoleAllocator() {
+        final int searchRow = Math.round(position.getPosY());
+        final int searchCol = Math.round(position.getPosX());
+
+        final int maxSearchRadius = this.worldSize / 2;
```
Having half the worldsize will lead to scenarios where they are at the edges and cant find anything right? Seems like that would cause a lot of frustration for the player.

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -188,12 +181,66 @@ public IRoleAllocator getRoleAllocator() {
+
+        final int maxSearchRadius = this.worldSize / 2;
+
+        final Collection<Optional<IStructure>> foundStructures = MatrixUtils.matrixSpiralSearch(
```
We only want one structure here right? So we want to return the first one we find aka the closest one

jonathan-lindqvist replied:
👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -188,12 +181,66 @@ public IRoleAllocator getRoleAllocator() {
+            maxSearchRadius,
+            1);
+
+        if (foundStructures.iterator().hasNext()) {
+            return foundStructures.iterator().next();
+        }
+        else {
+            return Optional.empty();
```
Question 
Im a bit lost here, mind explaining? :p

jonathan-lindqvist replied:
👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -188,12 +181,66 @@ public IRoleAllocator getRoleAllocator() {
+    }
+
+    @Override
+    public Collection<Optional<IStructure>> findNearestStructures(final Position position,
+                                                                  final int count) {
```
Also these methods are really similar, can't we use overloading (and just call the other repeatedly) or just have one that handles the case when we just want one.

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/inventory/IItem.java:
```diff
@@ -0,0 +1,8 @@
+package com.thebois.models.world.inventory;
+
+/**
+ * Temporary comment delete later.
+ */
+public interface IItem {
+
+}
```
You can now delete this and use the actual one 😄

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/inventory/package-info.java:
```diff
@@ -0,0 +1,5 @@
+/**
+ * Temporary delete later.
+ */
+
+package com.thebois.models.world.inventory;
```
Same with this!

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -11,12 +14,14 @@
 
     private Position position;
     private StructureType structureType;
+    private Collection<IItem> allNeededItems = new ArrayList<>();
+    private Collection<IItem> deliveredItems = new ArrayList<>();
```
Also `allNeededItems` is just for the blueprint, but will stay after the house is built, shouldn't it be a method along the lines of `isBlueprintDone()` that will check a `materialInventory` that it got everything it needs then convert it to the structure. The method could be called when someone drops something off to the inventory

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -11,12 +14,14 @@
      *
-     * @param posX          Position in X-axis
-     * @param posY          Position in Y-axis
+     * @param posX          Position in X-axis.
+     * @param posY          Position in Y-axis.
```
👏

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -61,4 +66,56 @@ public StructureType getType() {
+
+    @Override
+    public boolean deliverItem(final IItem deliveredItem) {
+        if (allNeededItems.contains(deliveredItem)) {
```
Inventories now have methods that allows us to check the amount of items in them! 😄

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -61,4 +66,56 @@ public StructureType getType() {
+    public float builtStatus() {
+        final float totalDelivered = this.deliveredItems.size();
+        final float totalNeeded = this.allNeededItems.size();
+        final float ratio = totalDelivered / totalNeeded;
```
Divide by 0? 👀

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -61,4 +66,56 @@ public StructureType getType() {
+        }
+    }
+
+    protected void setAllNeededItems(final Collection<IItem> allNeededItems) {
```
Agree, each structure should have a defined recipe from the start

JacobBredin replied:
Yup, has been removed

jonathan-lindqvist commented on src/main/java/com/thebois/utils/MatrixUtils.java:
```diff
@@ -27,4 +45,108 @@ private MatrixUtils() {
+     *
+     * @throws IllegalArgumentException Entered indices can't be negative
+     */
+    public static <TType> Optional<TType>[][] getSubMatrix(final Optional<TType>[][] matrix,
```
Might have missed or not gotten to that point but where is this used? And why do we need it?

jonathan-lindqvist commented on src/main/java/com/thebois/views/StructureView.java:
```diff
@@ -50,12 +50,17 @@ private void createHouseTexture() {
 
+    // Houses get different colors based on built status
+    private Color getHouseColor(final IStructure structure) {
+        return new Color(structure.builtStatus() / 2 + 1f / 2f, 0f, 0f, 1.0f);
```
Style
Doesnt matter but could be `getBuildStatus`, you decide 😄

jonathan-lindqvist reviewed (changes requested):
I like how the structures fade in when being built, gives the player feedback that something is happening!

Left some questions etc that you can look over! 👍

JacobBredin committed:
Merge branch 'master' into 12-build-using-resources

# Conflicts:
#	documents/diagrams/com.thebois.models.beings.jpg
#	documents/diagrams/com.thebois.models.world.jpg
#	documents/diagrams/com.thebois.utils.jpg
#	src/main/java/com/thebois/ColonyManagement.java
#	src/main/java/com/thebois/models/beings/AbstractBeing.java
#	src/main/java/com/thebois/models/beings/Colony.java
#	src/main/java/com/thebois/models/beings/Pawn.java
#	src/main/java/com/thebois/models/beings/README.md
#	src/main/java/com/thebois/models/world/World.java
#	src/main/java/com/thebois/utils/README.md
#	src/test/java/com/thebois/models/beings/BeingTests.java
#	src/test/java/com/thebois/models/world/WorldTests.java

JacobBredin committed:
Merge remote-tracking branch 'origin/master' into 12-build-using-resources

JacobBredin committed:
Add method isCompleted to IBuildable

JacobBredin committed:
Fix Pawn destination behavior to use pathfinder

JacobBredin committed:
Add usage of IItem in Pawn and AbstractStructure

JacobBredin committed:
Add tests for World and House

JacobBredin committed:
Rename and add a parameter to a method in Pawn

JacobBredin committed:
Add tests and comments

JacobBredin committed:
Add tests for House and AbstractStructure

martinjonsson01 commented on src/main/java/com/thebois/models/beings/AbstractBeing.java:
```diff
@@ -153,8 +155,48 @@ protected void setPath(final Collection<Position> path) {
+                firstLoop = false;
+            }
+
+            if (getPosition().manhattanDistanceTo(nearestPosition)
+                > getPosition().distanceTo(lastPosition)) {
```
question

How come you're comparing a manhattan distance to a euclidian distance? I don't see the reason for not using manhattan distance for both. And if there's a good reason, you should probably add a comment explaining that reason.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -22,19 +28,24 @@
      * @param destination   initial destination to travel to.
      * @param random        the generator of random numbers.
      * @param pathFinder    The generator of paths to positions in the world.
+     * @param finder          Finder used to locate things in the world
```
style

No point repeating the name/type of the parameter in the javadoc. Just "Used to locate things in the world." is enough to describe it.

Also, missing period.

martinjonsson01 replied:
And missing capitalization on two of the other parameter javadocs.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -22,19 +28,24 @@
     @Override
     public void update() {
         super.update();
+        updateClosestStructure();
```
style

You're not really updating the structure so much as you are *finding* the closest structure, right?

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
+        closestStructure.ifPresent(structure -> setFinalDestination(structure.getPosition()));
+    }
+
+    private void setFinalDestination(final Position destination) {
+        final Position nextTo = nearestPositionNextTo(destination);
+        if (getFinalDestination().isPresent()) {
+            if (!getFinalDestination().get().equals(nextTo)) {
+                final Collection<Position> newPath = getPathFinder().path(getPosition(), nextTo);
+                setPath(newPath);
+            }
+        }
+        else {
+            final Collection<Position> newPath = getPathFinder().path(getPosition(), nextTo);
+            setPath(newPath);
+        }
+    }
```
style

This isn't really a setter of the final destination as much as it is a calculator of a path to the final destination. So `calculatePathToDestination` is more accurate. Maybe breaking out the setting to outside the method is also appropriate in order to follow command/query separation.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
+        }
+    }
+
+    protected Optional<IStructure> findIncompleteStructure() {
+        final Collection<Optional<IStructure>> structures =
+            getFinder().findNearestStructures(this.getPosition(), 10);
+
+        for (final Optional<IStructure> structure : structures) {
+            if (structure.isPresent()) {
+                if (!structure.get().isCompleted()) {
+                    return structure;
+                }
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+        return Optional.empty();
+    }
```
This is the method that should be in the finder itself. You shouldn't ask the finder to get all nearby structures and then find the incomplete one yourself, you should ask the finder for the nearest incomplete structure directly, since that is the responsibility of the finder.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
+        if (closestStructure.isPresent() && getFinalDestination().isPresent()) {
+            // If the structure is completed or destination is unreachable pick a new one
+            if (closestStructure.get().isCompleted() || closestStructure.get().getPosition().equals(
+                getFinalDestination().get())) {
```
question

According to the comment,
```java
closestStructure.get().getPosition().equals(getFinalDestination().get())
```
being true means that "destination is unreachable". I don't see how the position of the closest structure being equal to the final destination of the pawn implies that this is the case. How come?

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
+                    return structure;
+                }
+            }
+            else {
+                return Optional.empty();
+            }
```
I think there's a logic error here, should it really be giving up and returning `empty()` if the first structure in the collection isn't present? I believe it should instead `continue` to allow the loop to search through the other structures in the list.

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
+        final Collection<Optional<IStructure>> structures =
+            getFinder().findNearestStructures(this.getPosition(), 10);
+
+        for (final Optional<IStructure> structure : structures) {
+            if (structure.isPresent()) {
+                if (!structure.get().isCompleted()) {
+                    return structure;
+                }
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+        return Optional.empty();
```
Can be simplified to:
```java
return structures
.toStream()
.flatMap(Optional::stream)
.filter(structure -> !structure.isCompleted())
.findFirst();
```

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +59,63 @@ private void setRandomDestination() {
+        }
+    }
+
+    protected void deliverItemToStructure(final Optional<IStructure> structure) {
```
style

How come there are `protected` methods in `Pawn`? I can't imagine any subtype of `Pawn` that would need access to these methods, so they should probably be private.

martinjonsson01 commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -262,7 +307,7 @@ public ITile getTileAt(final Position position) {
     @Override
     public ITile getTileAt(final int posX, final int posY) {
-        if (posX < 0 || posY < 0 || posX >= worldSize || posY >= worldSize) {
+        if (posX < 0 || posY < 0 || posX > worldSize || posY > worldSize) {
```
Since these positions are zero-indexed, one of them being equal to `worldSize` would mean that they are out of bounds. Why remove the `>=` checks?

martinjonsson01 replied:
If I try to get position (50, 50) in a world of size 50, then that would be out of bounds. (49, 49), however, is not.

martinjonsson01 replied:
Alright 👌 

Try not to commit bug-finding code in the future!

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -61,4 +66,62 @@ public StructureType getType() {
+    }
+
+    @Override
+    public boolean deliverItem(final IItem deliveredItem) {
```
style

I prefer to prefix this type of method with `try`, so it would be `tryDeliverItem`. Since it can fail if the item is not needed, what the caller is actually doing is *trying* to deliver an item.

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -61,4 +66,62 @@ public StructureType getType() {
+    }
+
+    @Override
+    public float builtStatus() {
```
style

Should be verb-ified to something like `getBuiltStatus` or `calculateBuiltStatus`.

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/IBuildable.java:
```diff
@@ -0,0 +1,50 @@
+    float builtStatus();
+
+    /**
+     * Returns true if builtStatus is equal to 1.
```
This is describing implementation details. The javadoc should describe what the method represents, not how it does it. So it should say "Returns true if construction is completed."

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -123,7 +135,12 @@ public void getPathReturnsEqualPositionsToPathFinder() {
+            endPosition,
+            mockRandom,
+            pathFinder,
+            new World(10, 0));
```
If the test doesn't need a real world, you probably shouldn't use a real world. Instead, create a mock of the interface that is being passed as an argument, e.g. `mock(IFinder.class)`. This also gives the benefit of tests not having to change when we update the constructor of world.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -249,4 +276,255 @@ public void equalsReturnsFalseForNotEqualBeings(
+    @Test
+    public void pawnBuildsHouses() {
+        // Arrange
+
```
style

Why the weird empty line?

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -249,4 +276,255 @@ public void equalsReturnsFalseForNotEqualBeings(
+        final float finalBuildStatus = structure.builtStatus();
+
+        // Assert
+
```
style

Are you putting these empty lines for a reason?

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -249,4 +276,255 @@ public void equalsReturnsFalseForNotEqualBeings(
+            world.getStructureAt(positionB).get().deliverItem(new Log());
+            world.getStructureAt(positionA).get().deliverItem(new Rock());
+            world.getStructureAt(positionA).get().deliverItem(new Log());
+
```
style

Weird empty line.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -249,4 +276,255 @@ public void equalsReturnsFalseForNotEqualBeings(
+        world.createStructure(positionA);
+        world.createStructure(positionB);
+
+        for (int i = 0; i < 10; i++) {
```
How does the test know that 10 of each item is required?

Looks like it knows it because it knows about the implementation details of house, and it knows that house internally says that it needs 10 of these items. 

This is incorrect, as the client code (currently this test) should have a real way of knowing for sure how many items are required. I believe this is possible using the current getters, but if it's not then this is a sign that you're missing a part of the required API.

JacobBredin replied:
The reason was to make sure no tests would run forever, which could be the case if it would check what would be needed by the house. Needing certain items to be constructed could be an requirement on a specific house, if so it is not out of place.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -249,4 +276,255 @@ public void equalsReturnsFalseForNotEqualBeings(
+        final Position correctDestination = new Position(0, 0);
+
+        final World world = new World(50, 0);
+        final IPathFinder pathFinder = new AstarPathFinder(world);
```
If you don't need the functionality of `AstarPathFinder` then you should mock it. Same goes for world, but in this case it looks like it is required.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -249,4 +276,255 @@ public void equalsReturnsFalseForNotEqualBeings(
+
+        final Random mockRandom = Mockito.mock(Random.class);
+        when(mockRandom.nextInt(anyInt())).thenReturn(0);
+        when(mockRandom.nextInt(anyInt())).thenReturn(0);
```
Duplicate line of code.

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -249,4 +238,167 @@ public void testsIfColonyGetsUpdate() {
         assertThat(world.getColony()).isEqualTo(colony);
     }
 
+    private boolean matrixEquals(final ITile[][] worldMatrix1, final ITile[][] worldMatrix2) {
+        for (int y = 0; y < worldMatrix2.length; y++) {
+            final ITile[] row = worldMatrix2[y];
+            for (int x = 0; x < row.length; x++) {
+                if (worldMatrix1[y][x] != null) {
+                    if (!worldMatrix1[y][x].equals(worldMatrix2[y][x])) {
+                        return false;
+                    }
+                }
+                else {
+                    if (worldMatrix2[y][x] != null) {
+                        return false;
+                    }
+                }
+            }
+        }
+        return true;
+    }
```
MatrixUtils?

martinjonsson01 reviewed (changes requested):
Left some questions and suggestions for the newly added parts!

JacobBredin committed:
Merge branch 'master' into 12-build-using-resources

# Conflicts:
#	documents/diagrams/com.thebois.models.beings.jpg
#	documents/diagrams/com.thebois.models.world.jpg
#	documents/diagrams/com.thebois.models.world.structures.jpg
#	documents/diagrams/com.thebois.utils.jpg
#	src/main/java/com/thebois/controllers/game/StructureController.java
#	src/main/java/com/thebois/models/beings/Colony.java
#	src/main/java/com/thebois/models/world/World.java
#	src/main/java/com/thebois/views/game/StructureView.java
#	src/test/java/com/thebois/models/beings/BeingTests.java
#	src/test/java/com/thebois/models/beings/ColonyTests.java
#	src/test/java/com/thebois/models/beings/roles/RoleAllocatorTests.java
#	src/test/java/com/thebois/models/world/WorldTests.java

JacobBredin committed:
Fix broken tests from the merge conflict

JacobBredin committed:
Make world return IStructures instead of copies

JacobBredin committed:
Add new colors to placed houses

JacobBredin committed:
Remove internal use of Optional in Pawn

JacobBredin committed:
Fix code style and naming in AbstractBeing

JacobBredin committed:
Add finder to constructor of AbstractBeing

JacobBredin committed:
Fix javadoc comment capitalization

JacobBredin committed:
Fix naming inconsistencies in AbstractBeing

JacobBredin committed:
Add findNearestIncompleteStructure to IFinder

JacobBredin committed:
Refactor Pawn behaviour

JacobBredin committed:
Change search area for structures to whole World

JacobBredin committed:
Remove unused IFinder method

JacobBredin committed:
Fix javadoc comments in IBuildable

JacobBredin committed:
Remove unused method findStructureAt

JacobBredin committed:
Add IFinder to colony constructor in Pawntastic

JacobBredin committed:
Change IBuildable to better conform with IInventory

JacobBredin committed:
Implement new IInventory methods in Colony

JacobBredin commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -61,4 +66,56 @@ public StructureType getType() {
+        this.allNeededItems = allNeededItems;
+    }
+
+    protected void setDeliveredItems(final Collection<IItem> deliveredItems) {
```
has been removed.

JacobBredin commented on src/main/java/com/thebois/models/world/structures/IBuildable.java:
```diff
@@ -0,0 +1,43 @@
+import com.thebois.models.world.inventory.IItem;
+
+/**
+ * IBuildable is an interface for interacting game entities that can be built.
```
Thanks, didn't think about it that way

JacobBredin committed:
Fix faults in javadoc comments and method names

JacobBredin commented on src/main/java/com/thebois/utils/MatrixUtils.java:
```diff
@@ -27,4 +45,108 @@ private MatrixUtils() {
+                try {
+                    subMatrix[row][col] = matrix[startRow + row][startCol + col];
+                }
+                catch (final ArrayIndexOutOfBoundsException exception) {
+                    subMatrix[row][col] = Optional.empty();
+                }
```
Don't know, method has been removed anyway, no need to worry 👍

JacobBredin committed:
Fix style faults in MatrixUtils and remove method

JacobBredin commented on src/main/java/com/thebois/models/beings/AbstractBeingGroup.java:
```diff
@@ -37,4 +40,12 @@ protected void setBeings(final Collection<IBeing> beings) {
+        return this.finder;
+    }
+
+    protected void setFinder(final IFinder finder) {
```
Has been moved to constructor 👍

JacobBredin committed:
Fix bug in Inventory calculateDifference

JacobBredin committed:
Rewrite and refactor unit tests for House

JacobBredin committed:
Remove unused condition from AbstractStructure

JacobBredin committed:
Add unit tests for Inventory and House

JacobBredin committed:
Add IInventory tests for Colony

JacobBredin committed:
Add full test coverage of IInventory in Colony

JacobBredin committed:
Remove equals and hash methods from Structure

Since AbstractStructure is a domain entity, it should not be equal to any instance except itself, and therefor only needs the standard equals and hash methods inherited from Object.

JacobBredin committed:
Add unit tests to Being and fix destination bug

Pawn now picks the final destination that is closest

JacobBredin committed:
Add unit tests to Being and Structure

JacobBredin committed:
Add folder test to CheckStyle suppressions.xml

JacobBredin committed:
Merge branch 'master' into 12-build-using-resources

# Conflicts:
#	src/main/java/com/thebois/models/world/World.java

JacobBredin committed:
Remove internal use of optionals from MatrixUtils

JacobBredin committed:
Fix javadoc comments and change method names

JacobBredin committed:
Remove unused import

JacobBredin committed:
Rename method names in WorldTests

JacobBredin committed:
Fix faulty style in unit tests

JacobBredin committed:
Rename MatrixToCollection to toCollection

JacobBredin committed:
refactor getHouseColor in StructureView

JacobBredin committed:
Rename unit tests in BeingTests

JacobBredin committed:
Refactor to use mock IFinder in tests

JacobBredin committed:
Rename unit tests in HouseTests

JacobBredin committed:
Remove duplicate line

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Colony.java:
```diff
@@ -25,7 +25,7 @@
 public class Colony extends AbstractBeingGroup implements IRoleAllocator, IInventory {
 
-    private final IInventory inventory = new Inventory();
+    private final IInventory colonyInventory = new Inventory();
```
style

It's an inventory in a colony. No need to name it `colonyInventory`, just `inventory` is fine.

JacobBredin replied:
it was hidden by the field in calculateDifference(final IInventory inventory)

JacobBredin replied:
Good idea, i've changed it

martinjonsson01 commented on src/main/java/com/thebois/models/beings/Colony.java:
```diff
@@ -45,8 +45,8 @@ public Colony(final Iterable<Position> vacantPositions, final IPathFinder pathFi
 
-    public IInventory getInventory() {
-        return inventory;
+    public IInventory getColonyInventory() {
```
style

Now you've made it into `colony.getColonyInventory()`, it should just be `colony.getInventory()`, no need to repeat the word "Colony" everywhere.

martinjonsson01 commented on src/main/java/com/thebois/models/inventory/IInventory.java:
```diff
@@ -77,4 +91,13 @@
      */
     int numberOf(ItemType itemType);
 
+    /**
+     * Calculates and returns a list of the differing IItems between two inventories.
+     *
+     * @param inventory The inventory to check against.
+     *
+     * @return A list of differing items.
+     */
+    ArrayList<ItemType> calculateDifference(IInventory inventory);
```
I can use this and an empty inventory to extract all of the item types in an inventory. That's a behavior that isn't really intended, but as they say, if something is public then someone somewhere will depend on it. 

Is there no other way to calculate the differing items? Could you not just use the counts of the inventories to calculate how far a structure has been built?

JacobBredin replied:
I need a way to compare two list. I don't see how that could be done without being able to check what's inside? But maybe there is a way around it?

JacobBredin replied:
Sure. I just read what you wrote about it further down, I could take a look if I can apply that idea

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -14,114 +15,84 @@
     private StructureType structureType;
-    private Collection<IItem> allNeededItems = new ArrayList<>();
-    private Collection<IItem> deliveredItems = new ArrayList<>();
+    private IInventory allNeededItems;
```
You could have this as just a list of item types required, so you don't need to compare the difference between two inventories.

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -14,114 +15,84 @@
-            return totalDelivered / totalNeeded;
-        }
+        return totalDelivered / totalNeeded;
     }
```
Are you assuming that neededItems is never empty? If so, there should probably a check for that in the constructor so that if someone tries to provide an empty list of needed items, it throws an InvalidArgumentException.

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -14,114 +15,84 @@
-
-    protected void setDeliveredItems(final Collection<IItem> deliveredItems) {
-        this.deliveredItems = deliveredItems;
+        return null;
```
Public interfaces should use optionals.

martinjonsson01 commented on src/main/java/com/thebois/views/game/StructureView.java:
```diff
@@ -42,7 +41,10 @@ public void draw(final Batch batch, final float offsetX, final float offsetY) {
     // Houses get different colors based on built status
     private Color getHouseColor(final IStructure structure) {
-        return Color.valueOf("1B52AB").lerp(Color.valueOf("3B2916"), structure.builtStatus());
+        final Color bluePrintColor = Color.valueOf("3B2916");
```
style

Blueprint is one word.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
         // Act
-        final float initialBuiltRatio = structureAtA.builtStatus();
-        testPawn.update();
+        pawn.update();
+        pawn.update();
```
question

Why the double-update?

JacobBredin replied:
It first needs to find an incomplete structure, then on the second update the pawn should look for a new closest structure since the current one isn't complete

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
 
         // Assert
-        assertThat(initialBuiltRatio).isLessThan(structureAtA.builtStatus());
+        final int callsExpected = 1;
```
This is fine to inline, don't need to break it out into a separate variable.

`times(1)` is actually more readable than `times(callsExpected)`

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
 
     @Test
-    public void findNewUnBuiltStructure() {
+    public void pawnTriesTpFindNewClosestStructureWhenNearestIsCompleted() {
```
typo

Tp -> To

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
 
-        final Random mockRandom = Mockito.mock(Random.class);
-        when(mockRandom.nextInt(anyInt())).thenReturn(0);
+        final IStructure mockStructureA = Mockito.mock(IStructure.class);
+        final IStructure mockStructureB = Mockito.mock(IStructure.class);
```
style

Instead of A and B, more meaningful names are `finishedStructure` and `unfinishedStructure`.

martinjonsson01 replied:
Alright, my suggested names don't fit then, but my point still stands. `xxxxxA` and `xxxxxB` are very non-meaningful names, which makes it harder to read the code and fully understand what is going on.

Since you know the code the best out of the two of us, you should be able to come up with better names for them.
Maybe something like `structureUnderConstruction` and `unfinishedStructure`?

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
+        pawn.update();
+
+        // Assert
+        final int callsExpected = 2;
+        verify(mockFinder, times(callsExpected)).findNearestIncompleteStructure(any());
```
style

Same here, `times(2)` is much more readable.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
+    }
+
+    @Test
+    public void pawnLooksForANewPawnWhenDestinationIsNotEqualToNeighborOfClosestStructure() {
```
style

Pawn looks for a new pawn?

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
+        final IPathFinder mockPathFinder = Mockito.mock(IPathFinder.class);
+        final IFinder mockFinder = Mockito.mock(IFinder.class);
+
+        final IStructure mockStructure = Mockito.mock(IStructure.class);
```
style

Could be called `unfinishedStructure` or `notCompletedStructure`

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
+        final IStructure mockStructure = Mockito.mock(IStructure.class);
+
+        final Position structurePosition = new Position(10f, 20f);
+        final Position positionA = new Position(0f, 0f);
+        final Position positionB = new Position(1f, 1f);
+        final Position positionC = new Position(2f, 2f);
+        final Position positionD = new Position(3f, 3f);
```
style

If these positions are not referenced anywhere, they'll be much less confusing if they're inlined into a list called `path`:
```java
List<Position> path = List.of(new Position(0f, 0f), [...]);
```
Otherwise the reader is wondering what A B C D means and what they're used for. If they have meaning outside of just being part of the path, you should try to give them more meaningful names.

martinjonsson01 commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -445,84 +440,117 @@ public void allStructuresCompletedNewPlaced() {
 
         // Assert
-        assertThat(initialBuiltRatio).isLessThan(structureAtA.builtStatus());
+        final int callsExpected = 3;
+        verify(mockPathFinder, times(callsExpected)).path(any(), any());
```
style

Could also be `times(3)`, if checkstyle doesn't complain?

martinjonsson01 commented on src/test/java/com/thebois/models/inventory/ColonyInventoryTests.java:
```diff
@@ -0,0 +1,124 @@
+                         new ArrayList<>(List.of(ItemType.ROCK))));
+    }
+
+    @ParameterizedTest
+    @MethodSource("calculatedDifferenceEqualToExpectedSource")
+    public void calculatedDifferenceEqualToExpected(
+        final Collection<IItem> listA,
+        final Collection<IItem> listB,
+        final ArrayList<ItemType> expectedDifference) {
+        // Arrange
+
+        final Stack<IItem> itemsA = new Stack<>();
+        final Stack<IItem> itemsB = new Stack<>();
+        itemsA.addAll(listA);
+        itemsB.addAll(listB);
+
+        final IInventory colonyA = new Colony(new ArrayList<>(),
+                                               Mockito.mock(IPathFinder.class),
+                                               Mockito.mock(IFinder.class));
+        final IInventory colonyB = new Colony(new ArrayList<>(),
+                                                          Mockito.mock(IPathFinder.class),
+                                                          Mockito.mock(IFinder.class));
+
+        colonyA.addMultiple(itemsA);
+        colonyB.addMultiple(itemsB);
+
+        // Act
+        final ArrayList<ItemType> difference = colonyA.calculateDifference(colonyB);
+
+        // Assert
+        assertThat(difference.containsAll(expectedDifference)).isTrue();
+    }
```
question

This is an inventory method, so it should be tested in `InventoryTests` right? Or why did you place it in a specific class for just `ColonyInventoryTests`?

martinjonsson01 commented on src/test/java/com/thebois/models/inventory/InventoryTests.java:
```diff
@@ -186,4 +190,85 @@ public void addMultipleItemsToInventory() {
+        final ArrayList<ItemType> expectedDifference) {
+        // Arrange
+
+        final Stack<IItem> itemsA = new Stack<>();
+        final Stack<IItem> itemsB = new Stack<>();
```
question

Why stacks?

JacobBredin replied:
I misread the method addMultiple in IInventory as using stack:  `addMultiple(final List<IItem> stack)`

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -42,7 +45,7 @@ public void getTypeEqualsHouse() {
 
     @Test
-    public void testEqualsHouse() {
+    public void areHousesEqual() {
```
style

Shouldn't be a question, should be an assertion, e.g. `getTypeReturnsHouse`.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -53,133 +56,131 @@ public void testEqualsHouse() {
         assertThat(structure.getType()).isEqualTo(StructureType.HOUSE);
```
style

Must've missed this before, but the act here is actually an arrange. And the `structure.getType()` is actuall the act. You're testing the `getType()` method in this test.

martinjonsson01 commented on src/test/java/com/thebois/utils/MatrixUtilsTests.java:
```diff
@@ -64,22 +61,18 @@ public void forEachElementGetsAllElementsInOrder(final Integer[][] matrix,
     @ParameterizedTest
     @MethodSource("getMatricesToTest")
-    public void matrixToCollectionTest(final Optional<String>[][] matrix,
+    public void matrixToCollection(final String[][] matrix,
```
style

You're just stating the name of the method you're testing. Describe what is being tested, e.g. `toCollectionTurnsMatrixIntoExpectedCollectionOfItems` or something similar.

martinjonsson01 commented on src/test/java/com/thebois/utils/MatrixUtilsTests.java:
```diff
@@ -64,22 +61,18 @@ public void forEachElementGetsAllElementsInOrder(final Integer[][] matrix,
-            { Optional.of("1"), Optional.empty(), Optional.of("3") },
-            { Optional.empty(), Optional.empty(), Optional.of("6") },
-            { Optional.of("7"), Optional.of("8"), Optional.of("9") },
+        final String[][] stringOptionalMatrix = new String[][] {
```
style

`stringOptionalMatrix` -> `stringMatrix`

And isn't there a better name? `stringMatrix` tells me nothing at all about what it represents. It's hard to parse out what it is supposed to contain without reading the entire file.

martinjonsson01 commented on src/test/java/com/thebois/utils/MatrixUtilsTests.java:
```diff
@@ -89,12 +82,9 @@ public void matrixToCollectionTest(final Optional<String>[][] matrix,
     @ParameterizedTest
     @MethodSource("getCorrectCoordinatesToTest")
-    public void testMatrixSpiralSearch(final Optional<String>[][] stringMatrix, final int startX,
+    public void matrixSpiralSearch(final String[][] stringMatrix, final int startX,
```
style

Don't just state the method name, describe how you're testing it.

martinjonsson01 reviewed (changes requested):
Great improvements, really close to merging now!

Just a few last finishing touches.

jonathan-lindqvist commented on src/main/java/com/thebois/models/inventory/IInventory.java:
```diff
@@ -77,4 +91,13 @@
+     *
+     * @param inventory The inventory to check against.
+     *
+     * @return A list of differing items.
```
ItemTypes (good) but the javadoc does not say that. Reading the javadoc I would have thought it would return IItems, might just be me though :p

jonathan-lindqvist replied:
👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -90,15 +91,6 @@ private void updateCanonicalMatrix() {
         return emptyPositions;
     }
 
-    /**
-     * Locates an object in the world and returns it.
-     *
-     * @return Object.
-     */
-    public Object find() {
-        return null;
-    }
-
```
🥳

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -120,15 +112,7 @@ public Object find() {
-            }
-        }
-        return copy;
+        return MatrixUtils.toCollection(this.structureMatrix);
```
Does this make them copies?

martinjonsson01 replied:
Don't think so, but they shouldn't be copies, right? Since a `Structure` is an entity with its own state, it's not a value object.

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -179,6 +163,50 @@ private void postObstacleEvent(final int posX, final int posY) {
+        final int searchRow = Math.round(position.getPosY());
+        final int searchCol = Math.round(position.getPosX());
+
+        final int maxSearchRadius = this.worldSize / 2;
```
Having half the worldsize will lead to scenarios where they are at the edges and cant find anything right? Seems like that would cause a lot of frustration for the player. (Instead of reopen so it gets into the new review)

jonathan-lindqvist replied:
![bild](https://user-images.githubusercontent.com/47867973/137345004-5479768a-1996-4268-be8c-0f13159ce8b9.png)
Like this

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -179,6 +163,50 @@ private void postObstacleEvent(final int posX, final int posY) {
+    }
+
+    @Override
+    public Optional<IStructure> findNearestIncompleteStructure(final Position position) {
+        final int searchRow = Math.round(position.getPosY());
+        final int searchCol = Math.round(position.getPosX());
+
+        final int maxSearchRadius = this.worldSize;
+
+        final Collection<IStructure> foundStructures =
+            MatrixUtils.matrixSpiralSearch(this.structureMatrix,
+                                           searchRow,
+                                           searchCol,
+                                           maxSearchRadius,
+                                           10);
+
+        for (final IStructure structure : foundStructures) {
+            if (!structure.isCompleted()) {
+                return Optional.of(structure);
+            }
+        }
+        return Optional.empty();
```
These to methods are almost identical, only difference are some values, make one method maybe?

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -11,27 +15,39 @@
 
     private Position position;
     private StructureType structureType;
+    private IInventory allNeededItems;
```
Think we talked about this being a hardcoded list of for each building? Then we would compare it that way, opinions?

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -11,27 +15,39 @@
+        final IInventory allNeededItems) {
         this.position = new Position(posX, posY);
         this.structureType = structureType;
+        this.allNeededItems = allNeededItems;
```
See comment about field, then we could skip this line 😄

JacobBredin replied:
There still needs to be a way to set what items are needed, this should be done by the structure factory

jonathan-lindqvist replied:
👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/world/structures/House.java:
```diff
@@ -10,20 +14,32 @@
+        super(position, StructureType.HOUSE, generateNeededItemsInventory());
+    }
+
+    private static IInventory generateNeededItemsInventory() {
+        final IInventory neededItems = new Inventory();
+
+        final int numberOfItems = 10;
+        for (int i = 0; i < numberOfItems; i++) {
+            neededItems.add(new Log());
+            neededItems.add(new Rock());
+        }
+
+        return neededItems;
```
Feels like a map is of better use in this scenario, would be something that would be like `RocksNeeded -> 10`

jonathan-lindqvist commented on src/test/java/com/thebois/models/inventory/ColonyInventoryTests.java:
```diff
@@ -0,0 +1,124 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+public class ColonyInventoryTests {
```
Weren't colony inventory tested before? Or did you just move it?

martinjonsson01 replied:
Looks like to me that only the new methods that were added to `IInventory` are tested in here. The rest are probably still tested in `ColonyTests`. Should probably be moved in here as well.

JacobBredin replied:
Since the method I added wasn't needed, ColonyInventoryTest has now been removed completely

jonathan-lindqvist reviewed (changes requested):
Super! 👏 
Though there are some questions and discussions to answer/resolve before this can be merged

JacobBredin committed:
Remove set and get for IFinder in BeingGroup

martinjonsson01 commented on src/main/java/com/thebois/models/beings/AbstractBeing.java:
```diff
@@ -153,8 +160,40 @@ protected void setPath(final Collection<Position> path) {
         return pathFinder;
     }
 
+    protected IFinder getFinder() {
+        return this.finder;
+    }
```
I think it's because fields should always be private, and they should be accessed through getters instead. As a subtype is a kind of client (albeit a more closely related one), you'd want the subtype to go through the getters as well.

martinjonsson01 replied:
On a related note (so I don't have to create a new review):

@JacobBredin you should probably rename this `IFinder` to be much more specific, since I've currently got a separate `IResourceFinder` on my branch. Maybe `IStructureFinder`.

JacobBredin replied:
Done!

JacobBredin committed:
Change parameter name in IInventory

JacobBredin commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -1,10 +1,15 @@
+import com.thebois.models.IFinder;
 import com.thebois.models.Position;
 import com.thebois.models.beings.pathfinding.IPathFinder;
+import com.thebois.models.inventory.items.Log;
+import com.thebois.models.inventory.items.Rock;
```
I know, but they cant collect material yet, so it is only a temporary solution

martinjonsson01 replied:
I wondered why it used these items, but the non line 85-86 you can see that they're used to deliver items to structures. This is fine for now, since I'll refactor it to be in a BuildTask later, so Pawn will not know about it then.

But if this branch has been merged with master, Log and Rock shouldn't exist anymore, you should have to use the factory to create the items (which means a looser dependency on just IItem).

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/Colony.java:
```diff
@@ -24,25 +25,28 @@
  */
 public class Colony extends AbstractBeingGroup implements IRoleAllocator, IInventory {
 
-    private final IInventory inventory = new Inventory();
+    private final IInventory colonyInventory = new Inventory();
```
👍

JacobBredin committed:
Refactor IFinder methods in  World

JacobBredin committed:
Refactor to use List instead of Stack in tests

JacobBredin committed:
Refactor getDifference to use Collections

JacobBredin committed:
Refactor calculateDifference to return Collection

JacobBredin committed:
Remove redundent methods from IInventory

JacobBredin committed:
Refactor IBuildable to use Optional

JacobBredin committed:
Refactor AbstractStructure to use EnumMap

Unused methods in IInventory has been removed and calculation of delivered items has been moved to AbstractStructure

JacobBredin committed:
Add unit test to StructureTests

JacobBredin committed:
Fix equals test in HouseTests

JacobBredin committed:
Fix style faults in test method names

JacobBredin committed:
Rename IFinder to IStructureFinder

JacobBredin committed:
Fix faulty method name in Beingtests

JacobBredin committed:
Add comments to Beingtests

JacobBredin committed:
Refactor and fix style errors in Beingtests

martinjonsson01 commented on src/main/java/com/thebois/models/beings/AbstractBeing.java:
```diff
@@ -169,7 +169,7 @@ protected IPathFinder getPathFinder() {
     }
 
-    protected IFinder getFinder() {
+    protected IStructureFinder getFinder() {
```
Rename to `getStructureFinder`

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/House.java:
```diff
@@ -30,14 +30,13 @@ public House(final Position position) {
 
-    private static IInventory generateNeededItemsInventory() {
-        final IInventory neededItems = new Inventory();
+    private static Map<ItemType, Integer> generateNeededItemsInventory() {
```
style

You're no longer generating an inventory, so maybe just something like `getNeededItems`.

JacobBredin replied:
Forgot to change it before I pushed it up, but I changed it to `getNeededItemsMap`

martinjonsson01 commented on src/main/java/com/thebois/models/world/structures/House.java:
```diff
@@ -30,14 +30,13 @@ public House(final Position position) {
-        }
+
+        neededItems.put(ItemType.LOG, numberOfItems);
+        neededItems.put(ItemType.ROCK, numberOfItems);
```
This looks a lot cleaner, nicely done!

martinjonsson01 commented on src/test/java/com/thebois/models/inventory/InventoryTests.java:
```diff
@@ -179,96 +179,12 @@ public void inventoryDoesNotHaveMultipleOfItem() {
 
         // Act
-        inventory.addMultiple(items);
+        inventory.addMultiple(List.of(new Log(), new Log()));
```
Looks like you haven't merged from master in a while.

After my item refactor that you'll get from master, you'll have to use `ItemFactory.createItem(ItemType.LOG)` instead.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -33,27 +34,30 @@ public void getPositionSetInConstructor() {
 
     @Test
-    public void areHousesEqual() {
+    public void housesAreEqualWhenSameInstance() {
```
Perfectly named tests! 😄

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -114,7 +118,7 @@ public void houseReturnsExpectedItemWhenDismantled(
         }
         // Act
-        final IItem retrieveItem = house.tryDismantle(itemTypeToRetrieve);
+        final IItem retrieveItem = house.tryDismantle(itemTypeToRetrieve).orElse(null);
```
You could use `.orElseThrow()` here to make the test fail if it can't dismantle the item.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -183,4 +187,37 @@ public void whenStructureHasReceivedItemsNeededItemsAreEqualToExpected(
         assertThat(neededItems.containsAll(expectedNeededItems)).isTrue();
     }
 
+    private static Stream<Arguments> whenItemsHaveBeenDeliveredStructureReturnsCorrectIsCompleteSource() {
```
style

I'm not a fan of this way of naming arguments. Couldn't it be called `getItemAmountsAndExpectedCompleteness` or something?

JacobBredin replied:
I named them like this because I noticed a lot of these "testArguments" methods would end up with the same names as each other even though they didn't have the same parameters or were for the same tests.

martinjonsson01 replied:
Do you have any examples? It sounds like they shouldn't have the same names if they don't have the same parameters or are not even used for the same tests.

JacobBredin replied:
Maybe I was just naming them wrong, guess it works

martinjonsson01 reviewed (changes requested):
Everything looks great! 

Except for the fact that you haven't merged with master in a while. You'll need to pull in the latest changes from master and fix the conflicts before this PR can be merged.

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -48,4 +65,26 @@ private void setRandomDestination() {
+        final Optional<IStructure> structure =
+            getFinder().findNearestIncompleteStructure(this.getPosition());
+        return structure.orElse(null);
+    }
```
👍

jonathan-lindqvist commented on config/suppressions.xml:
```diff
@@ -6,5 +6,6 @@
 
 <suppressions>
     <suppress checks="ClassDataAbstractionCoupling" files="Pawntastic"/>
+    <suppress checks="[a-zA-Z0-9]*" files="[\\/]test[\\/]"/>
```
🤖 👍

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/AbstractBeing.java:
```diff
@@ -153,8 +160,40 @@ protected void setPath(final Collection<Position> path) {
         return Optional.of(path.peek().deepClone());
     }
 
+    protected Position getFinalDestination() {
+        if (path.isEmpty()) return null;
+        return path.firstElement().deepClone();
+    }
+
```
Question

As usual I might be missing something obious, but what is the use case for this? Just wondering so that I get a better understand of the code 😄

JacobBredin replied:
AbstractBeing has another method called getDestination which is for getting the next position in the path, finalDestination is the absolut last position in the path. This will probably be moved to actions later

jonathan-lindqvist commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -14,28 +19,40 @@
     /* Temporary hard-coded world size. Should be removed when pathfinding is implemented. */
     private static final int WORLD_SIZE = 50;
     private final Random random;
+    private IStructure closestStructure;
```
`closestIncompleteStructure ` maybe?

jonathan-lindqvist reviewed (changes requested):
Some more questions an general feedback, so close to being able to merge! 💯 

Might want a review after merging with master/resolving conflicts, but that should be quick.

Great work!

JacobBredin committed:
Change variable and method names in Being

JacobBredin committed:
Refactor unit test in HouseTests

JacobBredin committed:
Merge branch 'master' into 12-build-using-resources

# Conflicts:
#	src/main/java/com/thebois/Pawntastic.java
#	src/main/java/com/thebois/models/inventory/items/Log.java
#	src/main/java/com/thebois/models/inventory/items/Rock.java
#	src/main/java/com/thebois/models/world/World.java
#	src/main/java/com/thebois/models/world/structures/AbstractStructure.java
#	src/main/java/com/thebois/models/world/structures/House.java
#	src/main/java/com/thebois/utils/MatrixUtils.java
#	src/main/java/com/thebois/views/game/StructureView.java
#	src/test/java/com/thebois/models/beings/BeingTests.java
#	src/test/java/com/thebois/models/beings/ColonyTests.java
#	src/test/java/com/thebois/models/inventory/InventoryTests.java
#	src/test/java/com/thebois/models/world/WorldTests.java
#	src/test/java/com/thebois/models/world/structures/HouseTests.java
#	src/test/java/com/thebois/models/world/structures/StructureTests.java
#	src/test/java/com/thebois/utils/MatrixUtilsTests.java

JacobBredin committed:
Fix master merge conflict

JacobBredin committed:
Fix bug in spiral search and add unit tests

Prevously implementation of matrixSpiralSearch traverse the matrix using a cross pattern, but this has been fixed to now use a correct spiral.

JacobBredin committed:
Add a unit test for House deepClone

JacobBredin committed:
Fix Item equals to only equal same ItemType

JacobBredin committed:
Decrease structure search radius in world to 10

JacobBredin committed:
Refactor findNearestIncompleteStructure in World

Search now uses a list of IStructures to find nearest incomplete structure instead of spiral search

JacobBredin committed:
Refactor findNearestIncompleteStructure

Code is now simpler and more straightforward

JacobBredin committed:
Remove all use of matrixSprialSearch from World

JacobBredin committed:
Merge branch 'master' into 12-build-using-resources

# Conflicts:
#	src/main/java/com/thebois/Pawntastic.java
#	src/main/java/com/thebois/models/world/World.java
#	src/main/java/com/thebois/models/world/structures/IStructure.java
#	src/test/java/com/thebois/models/beings/BeingTests.java

JacobBredin committed:
Make Item implement Serializable

JacobBredin committed:
Fix error in unit test parameter

JacobBredin committed:
Rename private method in House

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -291,4 +301,151 @@ public void getTerrainsReturnsTerrainEvenIfStructuresAreOnTopOfTerrain() {
         assertThat(actualNumberOfTerrainTiles).isEqualTo(expectedNumberOfTerrainTiles);
     }
 
+    private static Stream<Arguments> getCorrectCoordinatesToTest() {
```
style

This name is in the correct format, but also quite unclear. What is "correct"? And it's obvious that they are used "ToTest", we're in a test-class.

`getCoordinatesOfPositionAndClosestStructure` would fit.

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -291,4 +301,151 @@ public void getTerrainsReturnsTerrainEvenIfStructuresAreOnTopOfTerrain() {
+
+    }
+
+//    private static Stream<Arguments> getInCorrectCoordinatesToTest() {
+//        return Stream.of(Arguments.of(0, 0, 60, 60),
+//                         Arguments.of(5, 5, 100, 100),
+//                         Arguments.of(0, 0, 49, 26),
+//                         Arguments.of(20, 10, 49, 49)
+//        );
+//    }
+//
+//    @ParameterizedTest
+//    @MethodSource("getInCorrectCoordinatesToTest")
+//    public void findNearestStructureReturnsIncorrect(final int startX,
+//                                                     final int startY,
+//                                                     final int endX,
+//                                                     final int endY) {
+//
+//        // Arrange
+//        final World world = new TestWorld(50);
+//        world.createStructure(StructureType.HOUSE, endX, endY);
+//
+//        // Act
+//        final Optional<IStructure> foundStructure = world.findNearestStructure(new Position(
+//            startX,
+//            startY));
+//
+//        // Assert
+//        assertThat(foundStructure.isEmpty()).isTrue();
+//    }
```
Don't push up commented out code.

JacobBredin replied:
Fixed. Thought stylecheck would warn me :/

martinjonsson01 replied:
Yeah, we've got it so easy with it warning us about most things, so it's easy to forget the things it doesn't warn about

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -291,4 +301,151 @@ public void getTerrainsReturnsTerrainEvenIfStructuresAreOnTopOfTerrain() {
+//    }
+
+    @Test
+    public void returnsNoNearestStructure() {
```
style

When does it return no nearest structure?

`returnsNoNearestStructureWhenWorldIsEmpty` is more descriptive.

martinjonsson01 commented on src/test/java/com/thebois/models/world/WorldTests.java:
```diff
@@ -291,4 +301,151 @@ public void getTerrainsReturnsTerrainEvenIfStructuresAreOnTopOfTerrain() {
+        assertThat(structure.isPresent()).isFalse();
+    }
+
+    private static Stream<Arguments> getPositionsAndSizeToTest() {
```
Nicely named arguments-method!

martinjonsson01 replied:
You could do without the "ToTest"-suffix, though 😛

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -23,30 +34,206 @@ public void getPositionSetInConstructor() {
         final IStructure structure = StructureFactory.createStructure(
             StructureType.HOUSE,
             position);
+        // Act
+
```
style

Should be an empty line above, and no empty line below.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -23,30 +34,206 @@ public void getPositionSetInConstructor() {
+        assertThat(structureA).isNotEqualTo(structureB);
+    }
+
+    private static Stream<Arguments> tryDeliverItemReturnsExpectedValueSource() {
```
style

I can see why you couldn't think of a good name for this method, since it returns four different arguments grouped together. But still, it's not a good idea to just use the name of the test and append "Source". These argument-streams are supposed to be able to be used in multiple different tests.

JacobBredin replied:
I changed it, take a look

martinjonsson01 replied:
Much better, good job!

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/HouseTests.java:
```diff
@@ -23,30 +34,206 @@ public void getPositionSetInConstructor() {
+    @ParameterizedTest
+    @MethodSource("tryDeliverItemReturnsExpectedValueSource")
+    public void tryDeliverItemReturnsExpectedValue(
+        final int totalLogs,
+        final int totalRocks,
+        final IItem itemToDeliver,
+        final boolean expectedResult) {
```
Something to think about when using parameterized tests is to control your variables. And by that, I mean that you should only have as many free variables as you need to. In this case, you could have the delivered item be fixed, and only let the required amount of items vary. Or the other way around.

JacobBredin replied:
Thanks, I will look into it. But wouldn't that mean I would need many more tests that have houses with different amounts of items delivered to them? Right now I can tell the test what state the house should be in and what result I can expect from this state when an item is delivered.

martinjonsson01 replied:
Yeah, this way is for sure more compact and requires less tests.

The problem is that you're basically testing a lot of things at once, in a single test. So if this test fails, it has multiple reasons for why it could have failed. If it were split up into more specific cases, then when one of them fails you'll immediately be able to see what kind of functionality is not working.

martinjonsson01 replied:
Yeah, there are both upsides and downsides to splitting them up. If you don't want to spend time on refactoring the tests, that's fine since we're time-constrained this week.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/StructureFactoryTests.java:
```diff
@@ -28,7 +28,7 @@ public void factoryCreatesCorrectStructureWithGivenEnum(
 
         // Assert
-        assertThat(actualStructure).isEqualTo(expectedStructure);
+        assertThat(actualStructure.getClass()).isEqualTo(expectedStructure);
```
You could use `assertThat(x).isInstanceOf(Y.class)` here instead.

martinjonsson01 commented on src/test/java/com/thebois/models/world/structures/StructureTests.java:
```diff
@@ -14,7 +15,7 @@ public void structureDoesNotEqualNull() {
         final IStructure structure = StructureFactory.createStructure(StructureType.HOUSE, 0, 0);
 
         // Act
-        final boolean isEqual = structure.equals(null);
+        final boolean isEqual = structure == null;
 
         // Assert
         assertThat(isEqual).isFalse();
```
Here you're testing whether `createStructure` returns null or not.

So the `// Act` part is `createStructure` and the `// Assert` should be `assertThat(structure).isNotNull();`

martinjonsson01 replied:
But actually, isn't this test supposed to test the `.equals` method? Why did you change it to `==`?

martinjonsson01 reviewed (approved):
Left a few more suggestions and questions, and once they're all resolved I think you're done!

Oh, and it seems you've got an empty `Log.java` file, you should delete it.

martinjonsson01 commented on src/main/java/com/thebois/Pawntastic.java:
```diff
@@ -144,7 +144,7 @@ private void createModels() throws IOException, ClassNotFoundException {
             colony = new Colony(
                 world.findEmptyPositions(PAWN_POSITIONS),
-                new AstarPathFinder(world));
+                new AstarPathFinder(world), world);
```
Ah, now I know what it feels like to be on the other side of these things. I've gotten the exact same comment from Jonathan about the exact same thing but with IResourceFinder lol

JacobBredin commented on src/main/java/com/thebois/controllers/game/StructureController.java:
```diff
@@ -62,6 +62,7 @@ public IView getView() {
 
     @Override
     public void update() {
+        structureView.update(world.getStructures());
```
Seems like I accidentally added it because I didn't find where it was done when I changed the getStructures method.

Originally it was done on touch down, but I'm not sure it should? I can remove it, either way works.

JacobBredin replied:
Perhaps, depends on if we want to send a cloned list or not. Otherwise the View already has a reference to all the houses

JacobBredin commented on src/main/java/com/thebois/models/beings/Pawn.java:
```diff
@@ -14,28 +19,40 @@
+
+        if (closestIncompleteStructure != null) {
+            setFinalDestination(closestIncompleteStructure.getPosition());
+            deliverItemToStructure(closestIncompleteStructure);
```
Yeah that's more fitting. Good idea!

JacobBredin committed:
Rename method in Pawn

JacobBredin commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -213,6 +208,31 @@ private void postObstacleEvent(final int posX, final int posY) {
+
+    @Override
+    public Optional<IStructure> findNearestIncompleteStructure(final Position position) {
+        for (final IStructure structure : sortByDistanceTo(position)) {
+            if (!structure.isCompleted()) {
+                return Optional.of(structure);
+            }
+        }
+
+        return Optional.empty();
```
Thanks, the sorting I did was kind of unnecessary. Now it is cleaner and probably faster 👍

JacobBredin committed:
Refactor IStructureFinder methods in World

JacobBredin committed:
Refactor use of Map in AbstractStructure

JacobBredin commented on src/main/java/com/thebois/models/world/structures/AbstractStructure.java:
```diff
@@ -29,25 +43,60 @@ public Position getPosition() {
+    private int calculateSumOfAllNeededItems() {
+        int sum = 0;
+        for (final var entry : allNeededItems.entrySet()) {
+            sum += entry.getValue();
         }
-        final AbstractStructure that = (AbstractStructure) o;
-        return position.equals(that.position) && structureType == that.structureType;
+        return sum;
```
Does it become more readable though?

How about:
```
        int sum = 0;
        for (final Integer numberOfItems : allNeededItems.values()) {
            sum += numberOfItems;
        }
        return sum;
```

martinjonsson01 replied:
Sure, that's an improvement.

I think it might also be a factor of how used you are to reading stream-expressions.

What you basically do is just read it from left to right and ignore the `.stream()` bit. Then you can read the example above as "return all needed item values, then reduce them all into a single integer by summing them together".

I think it's worth getting used to them, as they can be extremely powerful when working with any form of collection, as you saw with my `getNearbyOfType`-example.

If you still don't want to use them, then you can go for the for-loop, that's fine.

martinjonsson01 replied:
They're relatively new, as they were introduced in Java 8, but they've since grown to be used quite widely, so definitely something [worth learning about](https://stackify.com/streams-guide-java-8/)!

(C# has a much cleaner version of the same concept, called [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/) - which I definitely recommend if you ever write something in C#!)

JacobBredin committed:
Fix javadoc style errors in MatrixUtils

JacobBredin committed:
Fix method style errors in HouseTests

JacobBredin committed:
Fix style errors in unit test method names

JacobBredin committed:
Refactor unit test for findNearestStructure

JacobBredin committed:
Fix unit test method names and style errors

JacobBredin commented on src/main/java/com/thebois/models/world/World.java:
```diff
@@ -139,13 +141,11 @@ public Object find() {
+        return structuresCache;
+    }
+
+    private void generateStructuresCache() {
+        structuresCache = MatrixUtils.toCollection(this.structureMatrix);
```
Instead of a split second lag-spike there is no lag spike, using the cache, which do you preferer?

martinjonsson01 replied:
That's what I was asking. Whether there was a noticable performance improvement to using this cache. And now you've confirmed that, yes, there is. 

And that's great! Then we should definitely keep using it, nicely done.

JacobBredin committed:
Remove duplicate update of StructuresView

JacobBredin committed:
Refactor countAllNeededItems to use stream

JacobBredin commented on src/test/java/com/thebois/models/beings/BeingTests.java:
```diff
@@ -256,17 +292,278 @@ public void equalsReturnsFalseForNotEqualBeings(
+    @ParameterizedTest
+    @MethodSource("pawnWalkToNearestUnBuiltStructureSource")
+    public void pawnWalkToNearestUnBuiltStructure(
+        final Position firstDestination, final Position secondDestination, final Position expectedPosition,
```
I refactored the test a bit, now it should be clearer. Now you have `startingPosition`, `expectedPosition` and `housePosition`

JacobBredin committed:
Refactor unit test in BeingTests

JacobBredin committed:
Add documentation to BeingTests

JacobBredin committed:
Rename method in WorldTests

Matte02 commented on src/main/java/com/thebois/models/inventory/items/Item.java:
```diff
@@ -21,4 +24,17 @@ public ItemType getType() {
     }
 
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null) return false;
+        final Item item = (Item) o;
+        return type == item.type;
```
Style
I'm pretty sure we said the only variable names that were allowed to be 1 char long is x and y.

Matte02 replied:
We have? Those I've written have used `object`, mostly because Jonathan pointed them out.

JacobBredin replied:
I changed all parameters named o to other

Matte02 reviewed (changes requested):
Well done!

Just 2 small things.

Matte02 commented on src/main/java/com/thebois/models/IStructureFinder.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public interface IStructureFinder {
+
+    /**
+     * Finds and returns the closest IStructure to the passed position.
+     *
+     * @param position The position to find a structure closest to.
+     *
+     * @return The found IStructure.
+     */
+    Optional<IStructure> findNearestStructure(Position position);
```
What do you mean you can't implement it yet? 
You have StructureType isn't that all you need to write the method?

Matte02 replied:
Exactly

martinjonsson01 replied:
That's what I did for `IResourceFinder`, and we should be consistent:

https://github.com/martinjonsson01/OOPP-WITH-THE-BOIS/blob/a3bbcac4785f01d9b0c67db71481b21d4e364b3f/src/main/java/com/thebois/models/world/World.java#L223

Though, now I'm seeing that mine probably shouldn't be called `getNearbyOfType`. We should agree on a common naming scheme for finders, what do you think about `getNearbyXXXOfType`? Or do you prefer `findNearestXXXOfType`?

Matte02 replied:
Now that you mention that, those methods are really similar, probably worth creating a helper method in order to not duplicate code in both methods.

martinjonsson01 replied:
They actually don't have any custom logic in them, they both already use the same utility methods (in the form of streams):
https://github.com/martinjonsson01/OOPP-WITH-THE-BOIS/blob/a3bbcac4785f01d9b0c67db71481b21d4e364b3f/src/main/java/com/thebois/models/world/World.java#L222-L229

JacobBredin replied:
I've added the parameter now 👍

jonathan-lindqvist commented on src/main/java/com/thebois/utils/MatrixUtils.java:
```diff
@@ -43,4 +64,49 @@ private MatrixUtils() {
+     *
+     * @return First found elements of type TType.
+     */
+    public static <TType> Collection<TType> matrixSpiralSearch(
+        final TType[][] matrix,
+        final int startRow,
+        final int startCol,
+        final int maxSearchRadius,
+        final int maxFoundElements) {
+
+        final Collection<TType> foundElements = new ArrayList<>();
+
+        for (int length = 0; length < maxSearchRadius; length++) {
+            for (int row = -length; row <= length * 2; row++) {
+
+                for (int col = -length; col <= length * 2; col++) {
+                    try {
+                        if (matrix[startRow + row][startCol + col] != null) {
+                            foundElements.add(matrix[startRow + row][startCol + col]);
+
+                            if (foundElements.size() > maxFoundElements) {
+                                return foundElements;
+                            }
+                        }
+                    }
+                    catch (final ArrayIndexOutOfBoundsException exception) {
+                        // Since there is a search limit, we can let the method loop
+                        // outside of matrix.
+                    }
+                }
+            }
+        }
+        return foundElements;
+    }
```
Saw some discussions about not using this? So is it still in use?

jonathan-lindqvist reviewed (approved):
Great work! 

Just some questions that need some answers and we will be good to go! 👍

JacobBredin committed:
Rename parameter o in equals methods to other

JacobBredin committed:
Add StructureType parameter to IStructureFinder

JacobBredin committed:
Fix CheckStyle error

Matte02 commented on src/main/java/com/thebois/models/IStructureFinder.java:
```diff
@@ -0,0 +1,33 @@
+     */
+    Optional<IStructure> findNearestStructure(Position position, StructureType type);
+
+    /**
+     * Finds and returns the closest incomplete IStructure to the passed position.
+     *
+     * @param position The position to find an incomplete structure closest to.
+     * @param type The type of structure to find.
+     *
+     * @return The found IStructure.
+     */
+    Optional<IStructure> findNearestIncompleteStructure(Position position, StructureType type);
```
Sorry if I'm annoying, but this method should not have a StructureType since builders don't care what structure to build, just that it's not completed.

JacobBredin replied:
I could see either way being useful, what if a specific role is tasked with only building a certain types of building? But then again, we would need a type for "any" building...

Matte02 replied:
Don't think you should have a type for all/any type. Or at least hope not, since that would break the structures button in my pr, since the buttons are based on the enum. 
Maybe have both methods if you can see a need for both of them.

JacobBredin replied:
I removed the StructureType from the method. If we need it, we can add it again later.

Matte02 reviewed (changes requested):
One more thing.

JacobBredin committed:
Rename IStructureFinder methods

JacobBredin committed:
Remove parameter from IStructureFinder method

Matte02 reviewed (approved):
Well done!

jonathan-lindqvist reviewed (approved):
LGTM

JacobBredin committed:
Remove unused method spiralSearch from utils
